Mathlib/Algebra/Ring/NonZeroDivisors.lean

```diff
@@ -44,6 +32,36 @@ theorem isRegular_iff_isUnit_of_finite {r : R} : IsRegular r ↔ IsUnit r where
 
 end Monoid
 
+section NonUnitalNonAssocRing
+
+variable {R : Type*} [NonUnitalNonAssocRing R] {r : R}
+
+lemma isLeftRegular_iff_eq_zero_of_mul_left : IsLeftRegular r ↔ ∀ x, r * x = 0 → x = 0 where
+  mp h r' eq := h (by simp_rw [eq, mul_zero])
+  mpr h r₁ r₂ eq := sub_eq_zero.mp <| h _ <| by simp_rw [mul_sub, eq, sub_self]
+
+lemma isRightRegular_iff_eq_zero_of_mul_right : IsRightRegular r ↔ ∀ x, x * r = 0 → x = 0 where
+  mp h r' eq := h (by simp_rw [eq, zero_mul])
+  mpr h r₁ r₂ eq := sub_eq_zero.mp <| h _ <| by simp_rw [sub_mul, eq, sub_self]
+
+lemma isRegular_iff_eq_zero_of_mul :
+    IsRegular r ↔ (∀ x, r * x = 0 → x = 0) ∧ (∀ x, x * r = 0 → x = 0) := by
+  rw [isRegular_iff, isLeftRegular_iff_eq_zero_of_mul_left, isRightRegular_iff_eq_zero_of_mul_right]
+
+/-- A (not necessarily unital or associative) ring has no zero divisors
+iff multiplication is both left and right cancellative. -/
+lemma noZeroDivisors_tfae : List.TFAE
+    [NoZeroDivisors R, IsLeftCancelMulZero R, IsRightCancelMulZero R, IsCancelMulZero R] := by
```

Comment:
Given the docstring, I think this might be better as two statements,
* `NoZeroDivisors R ↔ IsCancelMulZero R` (which is what the docstring says)
* `List.TFAE [IsLeftCancelMulZero R, IsRightCancelMulZero R, IsCancelMulZero R]` (which is the thing that is hard to write documentation for)